### DISCUSSION TITLE
A kind of "-normalize" that only looks at a part of an image?

### DISCUSSION THREAD

**Discussion by kamocio** [ImageMagick/ImageMagick#2610] [Help]
Hi, 

I am processing a number of images taken by an automated roller coaster ride capture system. Being an outdoor location, the lighting is very variable throughout the day, and depending on natural light. 

The capture DSLR is fixed, and there is a part of the image that contains subjects, pretty much always in the same spot.

There is also, unfortunately, a good portion of the sky that can be very bright, so, our trusty "-normalize" flag does no good. Now, we don't care if the sky over-saturates - all we care about is that the subject's face is has more or less proper brightness. 

Hence, the question: is it possible to run a command that does to the entire image what "-normalize" would do, but limiting the "evaluation" of histogram only to a certain region of the image, so that super bright areas we dont care about don't  end up capping the histogram stretching?

**Comment by snibgo** [comment]
You can crop to your area of interest, and normalize that. But IM has no facility to tell you the equivalent "-level" command that you could then apply to the entire image.

If you build IM yourself, my process modules can be used to do the histogram work. For example, suppose we want to make 2% black and 3% white, Windows CMD syntax:

```
%IM7DEV%magick ^
  toes.png ^
  -crop 100x100+100+100 +repage ^
  -process 'mkhisto cumul norm' ^
  -process invclut ^
  ( -clone 0 -crop 1x1+%[fx:w*0.02]+0 +repage ) ^
  ( -clone 0 -crop 1x1+%[fx:w*0.97]+0 +repage ) ^
  -delete 0 ^
  +append +repage ^
  txt:
```
The text output is:
```
# ImageMagick pixel enumeration: 2,1,4294967295,srgb
0,0: (1.28896e+09,1.09707e+09,1.0069e+09)  #4CD441643C04  srgb(30.011%,25.5432%,23.4436%)
1,0: (3.72605e+09,3.02881e+09,3.03111e+09)  #DE16B487B4AA  srgb(86.7538%,70.52%,70.5734%)
```
The percentage numbers can be used for "-level" in each channel. If we want the same "-level" for all channels, just pick one from each list, eg "-level 23.44,70.52%.

An alternative method is to compare versions of the crop, before and after the "-normalize". By choosing corresponding pixels from both versions, provided they haven't clipped, we can calculate the "-level" parameters. See my replies in thread https://github.com/ImageMagick/ImageMagick/discussions/2383

**Comment by fmw42** [comment]
If on IM 6, you could try using -region to process only that rectangular area

`convert image -region WxH+X+Y -normalize +region result`

-region may not work the same in IM 7, but you can try and see what happens.

Alternately, make a mask of the region you want to process. Then process the whole image. Then do a composite of the original with the processed image using the mask to control the blending.  

`convert image processed mask -compose over -composite result`

In IM 7, use magick rather than convert.

**Reply by kamocio** [reply]
Very interesting, thanks! 

I'm on IM v 7.0.5-4, and indded, +region does not seem to work. E.g.

`convert 1.jpg -region 10x10+620+900 -normalize +region out.jpg`

gives an error: `InvalidArgument '+region' '(null)' at CLI arg 5 @ error/operation.c/CLISimpleOperatorImage/3109`

Re: masking the region, not sure what to use instead of names in the command line you suggested (e.g. what are the "processed" and "mask" in

`convert` image processed mask -compose over -composite result`

?

**Reply by fmw42** [reply]
Is 10x10 your dimensions or offset. Seems like a very small region.  -region WxH+X+y where W=width, H=height, X=xoffset from top left and Y=yoffset. As I said IM 7 behaves differently and may not work. But note that on IM 7, use magick in place of convert.

image is input image
processed is the name of the file you create via `convert image -normalize processed`
mask is the name of the binary image that is white where you want to see the normalized result and black where you want to see the original.  

But note that the whole image will be processed from the statistics of the full image. You will only see the normalized pixels where the mask is white in the output.

-region should only process that region.

**Reply by kamocio** [reply]
10x10 would be the dimension of the region to sample. 

I tried the same cmd line with IM 7.0.10 - and while the command itself (e.g. `magick in.jpg -region 100x100+0+0 -normalize +region out.jpg`) executes (no errors) - the result is that only the specified 100x100 region is normalized, and I end up with a normalized square in the middle of the image :-) In other words, the region is not used just for sampling and determining normalization level, which then is applied to the whole image.

**Reply by fmw42** [reply]
That is what I tried to tell you earlier. -region only processes that region.  What is your platform? If unix-based, then I could script something most likely.

**Reply by kamocio** [reply]
Thanks! Really appreciate it. We are a windows shop, unfortunately.
Maybe if there was a way to tell levels of an image, we could crop a region of interest out, analyze it to get the levels and run -sigmoidal-contrast or smth similar with a parameter, but as per @snibgo , that's not possible either.

**Comment by snibgo** [comment]
I have shown you how to do this.

A simpler alternative (which doesn't do exactly what you want) is to apply the "-level" to the entire image that has the same effect that "-auto-level" would give to a selected crop. It works by finding the minimum and maximum values in the crop, and using those directly in "-level". Windows BAT syntax:
```
magick ^
  in.png ^
  ( +clone ^
    -crop 10x10+620+900 +repage ^
    -set option:CMIN %%[fx:minima*QuantumRange] ^
    -set option:CMAX %%[fx:maxima*QuantumRange] ^
    +delete ^
  ) ^
  -level %%[CMIN],%%[CMAX] ^
  out.png
```

**Reply by kamocio** [reply]
@snibgo , thank you so much, and sorry I was not getting this. I am new to this and those commands look alien to me. E.g. this "fx:minima*QuantumRange" 

I love the script though! Very fast, and got potential. I'm not entirely sure how it works, but it kind-a looks like it does what I want ... Its not as predictable as -normalize, e.g. if I specify a white region - the whole image turns white and white turns black, and if I specify a region which is definitely underexposed - it does bump up the brightness, but just barely. I do fx:maxima*QuantumRange*0.9, and its better, but not even sure what it does.

Again, thank you so much for the help.

**Comment by fmw42** [comment]
Create a binary mask where the region you want to change is white and the rest is black.  Normalize the image as per what snibgo provided. Then use the mask to composite the original and the normalized image.

`convert original normalized mask -compose over -composite result.`

You can even anti-alias the transition in the mask

`convert original normalized \( mask -blur 0x2 -level 50x100% \) -compose over -composite result`

The above is unix syntax. If on Windows, then remove the \. If in a bat file, then double the % to %%
